cpp/arcticdb/version/symbol_list.cpp

```diff
@@ -6,292 +6,703 @@
-        std::vector<StreamId> stream_ids;
-        store->iterate_type(KeyType::VERSION_REF, [&] (const auto& key) {
-            auto id = variant_key_id(key);
-            stream_ids.push_back(id);
-
-            if (stream_ids.size() == this->max_delta_ * 2 && !warned_expected_slowdown_) {
-                log::version().warn(
-                        "No compacted symbol list cache found. "
-                        "`list_symbols` may take longer than expected. \n\n"
-                        "See here for more information: https://docs.arcticdb.io/technical/on_disk_storage/#symbol-list-caching\n\n"
-                        "To resolve, run `list_symbols` through to completion to compact the symbol list cache.\n"
-                        "Note: This warning will only appear once.\n");
-
-                warned_expected_slowdown_ = true;
-            }
-        });
-        auto res = batch_get_latest_version(store, version_map_, stream_ids, false);
+    return output;
+}
 
-        CollectionType symbols{};
-        for(const auto& map_pair: *res) {
-            symbols.insert(map_pair.first);
-        }
+std::vector<SymbolListEntry> read_from_storage(
+        const std::shared_ptr<StreamSource>& store,
+        const AtomKey& key) {
+    ARCTICDB_DEBUG(log::symbol(), "Reading list from storage with key {}", key);
+    auto [_, seg] = store->read_sync(key);
+    if(seg.row_count() == 0)
+        return {};
+
+    missing_data::check<ErrorCode::E_UNREADABLE_SYMBOL_LIST>( seg.descriptor().field_count() > 0,
+                                                              "Expected at least one column in symbol list with key {}", key);
+
+    if(seg.descriptor().field_count() == 1)
+        return read_old_style_list_from_storage(seg);
+    else
+        return read_new_style_list_from_storage(seg);
+}
+
+bool is_new_style_key(const AtomKey& key) {
+    return util::variant_match(key.end_index(),
+    [] (std::string_view str) {
+        return str == version_string;
+    },
+    [] (NumericIndex n) {
+        return n == version_identifier;
+    });
+}
+
+MapType load_journal_keys(const std::vector<AtomKey>& keys) {
+    MapType map;
+    for(const auto& key : keys) {
+        const auto& action_id = key.id();
+        if(action_id == compaction_id)
+            continue;
+
+        const auto& symbol = key.start_index();
+        const auto version_id = is_new_style_key(key) ? key.version_id() : unknown_version_id;
+        const auto timestamp = key.creation_ts();
+        ActionType action = std::get<StringId>(action_id) == DeleteSymbol ? ActionType::DELETE : ActionType::ADD;
+        map[symbol].emplace_back(version_id, timestamp, action);
+    }
+    return map;
+}
+
+auto tail_range(const std::vector<SymbolEntryData>& updated) {
+    auto it = updated.crbegin();
+    const auto reference_id = it->reference_id_;
+    ++it;
+
+    while(it != updated.crbegin() && it->reference_id_ == reference_id)
```

Comment:
The list is sorted by reference id, so we're sorting backwards from the most recent looking for things that have the same reference id (i.e. conflicts). We don't care about conflicts that aren't the most recent as we assume that they will have been superseded by what happens after - for example if I race a write and delete on version 2, then have a non-racy write on version 3 then there is no problem because the symbol with both exist and be in the list.